### DISCUSSION TITLE
pip-tools について調査

### DISCUSSION THREAD

**Discussion by habvi** [42-pong/42-pong#61] [backend]
- #24 より、requirements.txt を生成するコマンド `pip-compile` を使う際に install する `pip-tools` について調べてみる

pip-tools
https://github.com/jazzband/pip-tools

**Comment by habvi** [comment]
この図が分かりやすかったです
[pip-tools の github](https://github.com/jazzband/pip-tools) より。

![image](https://github.com/user-attachments/assets/d6b69f2d-7412-4c68-a054-c8aef3c35260)


<br>

調べてみての個人的な結論はこんな感じです
- BE コンテナ build 時は今のまま `pip install -r requirements.txt` で良い
- BE コンテナを起動したまま `requirements.txt` と同期したい場合や、ローカルで venv など仮想環境を使っている場合は `pip-sync` がより良い
- `requirements.in` も共有し、Makefile で楽に `requirementx.txt` を生成できるルールを追加する

<br>

以下、調べたこと

### 各ファイル・コマンドの役割・挙動
- `requirements.in`
  - 役割
    - パッケージをリストアップするため (だけ) のファイル
  - `pip install -r requirements.in` すると
    - リストにのっているパッケージを install するだけ。リストにのっていない・install 済みのパッケージは uninstall されない
    - version を書いてないものは自動で最新になったり、パッケージ同士の依存関係が変わった場合に環境の再現性が保証されない
- `requirements.txt`
  - 役割
    - `requirements.in` を元に `pip-compile` で生成された、依存関係が固定されたリスト
  - `pip install -r requirements.txt` すると
    - リストにのっているパッケージを install するだけ。リストにのっていない・install 済みのパッケージは uninstall されない
    - `requirements.in` から生成された `requirements.txt` であれば、リストにのっているパッケージの部分の再現性は保証される

<br>

pip-tools の 2 つのコマンド
- `pip-compile`
  - `requirements.in` から、全ての依存関係とその依存関係の version が固定された `requirements.txt` のリストを生成する 
- `pip-sync`
  - `requirements.txt` にのっているパッケージを install し、のっていないパッケージを uninstall する (同期する)
  - 仮想環境を `requirements.txt` の内容と一致させることにより、環境の再現性が保証される

<br>

まとめると、
- `requirements.in` : 依存関係解析は特にしていない、install したいパッケージのリスト
- `requirements.txt` : `pip-compile requirements.in` により依存関係解析後の、全てのパッケージの version が固定されたリスト。固定された環境の共有に使用する
- `pip install -r requirements.txt` : `requirements.txt` にのっているパッケージを install する
- `pip-sync` : `requirements.txt` にのっているパッケージを install し、のっていないパッケージを uninstall するため、仮想環境内のパッケージが同期される

という理解をしました。
何かあればお願いします！🙏

**Reply by ayase-mstk** [reply]
大変わかりやすい資料をありがとうございます！

ローカルのコンテナとの環境同期にpip-sync使えるの初めて知りました！
